LMDB fairer lock synchronisation

wotbrew opened the pull request:
This PR _should_ fix #1826

## Problem

Release 1.22.0 introduced an additional write lock to avoid segfault when handling `MDB_MAP_FULL` errors (#1816). The write lock was used around all writes, as exclusivity of write transaction is necessary anyway for LMDB. This had an unintended consequence where the `acquire-write-lock` timeout was now used for ingester writes, where it would previously only be used for resizes.

Now, the timeout itself is fairly generous for writes, however due to the lack of fairness/write priority with the StampedLock mechanism, @tatut described running into ingester timeouts when ingesting nodes were also under high concurrent read load.

## Fix

This PR introduces a new locking mechanism in the java class `xtdb.lmdb.MapResizeSync`. It provides shared locks for open transactions, and exclusive 'resize' lock to deal with `MDB_MAP_FULL`.  It uses the base `AbstractQueueSynchronizer` to provide fairness and write priority to avoid in more situations reads blocking the ingester.

## Write lock timeouts

The PR as it stands _removes the write lock timeout_.  

I do not think there is a single good policy here, it depends ultimately on your write workload, but also how long at the limit your queries take, which is presumably open for many deployments.

I would welcome comments, not sure if a chesterton's fence situation. I understand that the timeout was introduced with #945 to provide observability of deadlock or otherwise blocking for too long. I think there may be better ways to achieve this without failing the ingester during resize.

I propose instead of timing out, we implement an ingester heartbeat (if it does not already exist) which will `WARN` if the ingester is blocked (for any reason, not just on lock acquisition). Perhaps as part of #1831. There is also deadlock/contention detection `ThreadMXBean` we can leverage to provide telemetry more holistically.

wotbrew committed:
Remove redundant max-readers check, mistakenly checks a var rather than opt

nat-int required for env-maxreaders anyway so will always fire - Therfore removing.

wotbrew committed:
Use a custom queue synchronization implementation (MapResizeSync) to provide
write preference, fairness and barging under contention.

Removed lock timeout error for now - pending consultation to figure out correct policy.

jarohen commented on modules/lmdb/project.clj:
```diff
@@ -11,6 +11,13 @@
 
   :scm {:dir "../.."}
 
+  :java-source-paths ["src"]
```
you'll need to pull in the javadoc and sources classifiers from another module btw (e.g. core) - now that you've got a public Java class Maven Central will expect javadoc and sources jars

wotbrew committed:
Change MapResizeSync fairness policy due to deadlock possibility.

It was possible that nested reads would race with a queued writer leading to deadlock. Solution is to not queue (park) writes, instead spin. This provides fairness between readers, writers due to barging should often win over queued readers though there is no longer write fairness.

Map resizes are infrequent events, I believe this policy will work in practice.

wotbrew committed:
Include :java-doc-opts and :classifiers in lmdb due to new java sources
